### PR TITLE
Solved crypto square exercise

### PR DESCRIPTION
**Pull request by iuliagatea**

### CODE REVIEW

File: Ruby/crypto-square/crypto_square.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  private
+
+  attr_reader :plaintext
+
+  def normalize
+    plaintext.downcase.delete(' ').gsub(/[^0-9A-Za-z ]/, '')
+  end
+
+  def number_of_rectangle_rows
+    (normalize.length**0.5).ceil
```
Comment (paucio) [inline Ruby/crypto-square/crypto_square.rb]: `normalize` and `number_of_rectangle_rows ` are another example where you need to memoize the values.

File: Ruby/crypto-square/crypto_square.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
+class Crypto
+
+  def initialize(plaintext)
+    @plaintext = plaintext
+  end
+
+  def ciphertext
+    plaintext.empty? ? '' : rectangle.transpose.map(&:join).join(' ')
```
Comment (paucio) [inline Ruby/crypto-square/crypto_square.rb]: I would use here a guard clause and extract  `rectangle.transpose.map(&:join).join(' ')` into a private method.

https://medium.com/the-code-review/en-garde-how-you-can-refactor-nested-ifs-with-guard-clauses-in-ruby-1ff728b7a0cd

File: Ruby/crypto-square/crypto_square.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    plaintext.downcase.delete(' ').gsub(/[^0-9A-Za-z ]/, '')
+  end
+
+  def number_of_rectangle_rows
+    (normalize.length**0.5).ceil
+  end
+
+  def rectangle
+    rows = normalize.chars.each_slice(number_of_rectangle_rows).to_a
+    rows.last << ' ' while rows.last.length < number_of_rectangle_rows
```
Comment (paucio) [inline Ruby/crypto-square/crypto_square.rb]: I'm not sure what are you doing here... Could you explain?

Comment (iuliagatea) [reply Ruby/crypto-square/crypto_square.rb]: i slice the text into `number_of_rows` rows and then i add empty spaces to the last element, if the length of `last_element` is smaller than the other rows

Comment (paucio) [reply Ruby/crypto-square/crypto_square.rb]: I understand now... Thank you 👍

File: Ruby/crypto-square/crypto_square.rb (hunk 1 above)
Comment (iuliagatea) [reply Ruby/crypto-square/crypto_square.rb]: done

File: Ruby/crypto-square/crypto_square.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  def ciphertext
+    plaintext.empty? ? '' : rectangle.transpose.map(&:join).join(' ')
+  end
+
+  private
+
+  attr_reader :plaintext
+
+  def normalize
+    plaintext.downcase.delete(' ').gsub(/[^0-9A-Za-z ]/, '')
```
[the remark this answers is not in the document]
Comment (iuliagatea) [reply Ruby/crypto-square/crypto_square.rb]: done

File: Ruby/crypto-square/crypto_square.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  end
+
+  def rectangle
+    rows = normalize.chars.each_slice(number_of_rectangle_rows).to_a
+    rows.last << ' ' while rows.last.length < number_of_rectangle_rows
+    rows
+  end
+
+  def encode
+    rectangle.transpose.map(&:join).join(' ')
```
Comment (asarbu) [inline Ruby/crypto-square/crypto_square.rb]: `rectangle` is not a very suggestive name (I know this comes from the text, but I don't think this is part of the domain)

File: Ruby/crypto-square/crypto_square.rb [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  def normalize
+    @normalize ||= plaintext.downcase.delete(' ').gsub(/[^\dA-Za-z]/, '')
+  end
+
+  def number_of_rectangle_rows
+    @number_of_rectangle_rows ||= (normalize.length**0.5).ceil
+  end
+
+  def rectangle
+    rows = normalize.chars.each_slice(number_of_rectangle_rows).to_a
```
Comment (asarbu) [inline Ruby/crypto-square/crypto_square.rb]: why not make `normalize` an action that changes your input ? This way in `encode` you could do smth like:
```ruby
def encode
  normalize!
  new_order = reposition_letters
  transpose!(new_order)
end
```

Comment (iuliagatea) [reply Ruby/crypto-square/crypto_square.rb]: why should it change the input? this is the best practice? i don't need normalize anywhere else

Comment (asarbu) [reply Ruby/crypto-square/crypto_square.rb]: `normalize` is a verb => it should do something

`normalized_input` is a noun => it should return something (input in a normalized state)

Comment (iuliagatea) [reply Ruby/crypto-square/crypto_square.rb]: I pushed a new commit

**Pull request merged**